events/controller.go

```diff
@@ -0,0 +1,63 @@
+package events
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/awslabs/operatorpkg/object"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Controller[T client.Object] struct {
+	kind       string
+	startTime  time.Time
+	kubeClient client.Client
+}
+
+func NewController[T client.Object](client client.Client, clock clock.Clock) *Controller[T] {
+	return &Controller[T]{
+		kind:       reflect.TypeOf(object.New[T]()).Elem().Name(),
+		startTime:  clock.Now(),
+		kubeClient: client,
+	}
+}
+
+func (c *Controller[T]) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		For(&v1.Event{}, builder.WithPredicates(predicate.NewTypedPredicateFuncs(func(o client.Object) bool {
```

Comment:
Something I don't know: Can events be updated? If so, how do we handle that so that we don't double-fire for events?

Reply:
Seems like an event can be updated/patched https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/#Event. An updated predicate should resolve that, no?

Reply:
Right -- I really don't think that it's likely that these events get updated/patched though -- it's probably fine to not handle this and only deal with this if it actually becomes a problem